<a href="https://colab.research.google.com/github/erestrepol/Business-Intelligence/blob/main/Reto_2_%C3%81rboles_de_Decisi%C3%B3n_(Aprendizaje_Supervisado)_ERL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se procede con la carga de las librerías de trabajo.

In [ ]:
#Librerias cientificas
import numpy as np
import pandas as pd

#Librerias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

 #Conectar al drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Se procede con la carga de los datos

In [ ]:
nxl='/content/drive/MyDrive/2. CollegeAdmisions_Int_M.xlsx'
XDB= pd.read_excel(nxl,sheet_name=0) #Base AAA
XDB.head(100)
XDB=XDB.dropna() #Eliminar filas vacias
XDB.head(100) #Mostrar 100 lineas del archivo

#Seleccionar las variables
XD=XDB[['Edu_Parent1', 'Edu_Parent2', 'Gender', 'White', 'Asian','HSGPA', 'SAT/ACT']]
XD.loc[:,'Gender'] = XD['Gender'].map({'F': 0, 'M': 1})
yd=XDB[['Admitted']]

Edu_Parent1  Edu_Parent2 Gender  White  Asian  HSGPA  SAT/ACT
6                6            6      0      1      0   4.26     1430
28               7            6      1      1      0   3.95     1340
46               3            6      1      1      0   4.33     1200
61               7            7      0      1      0   4.27     1530
65               6            7      0      0      0   4.04     1270
...            ...          ...    ...    ...    ...    ...      ...
17330            2            5      1      0      0   2.80     1010
17332            7            4      1      1      0   3.04     1130
17333            4            6      1      1      0   2.80     1210
17334            7            6      1      0      0   3.42     1000
17337            7            6      1      1      0   3.05     1130

[1528 rows x 7 columns]

2. Se procede con la implementación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD,yd)

#Se procede con la gráfica
from pydotplus import graph_from_dot_data #Crea la gráfica
from sklearn.tree import export_graphviz #Lo exporta a PNG, JPEG, TIFF

ve=['Edu_Parent1', 'Edu_Parent2', 'Gender', 'White', 'Asian','HSGPA', 'SAT/ACT']
dot_data=export_graphviz(mar, feature_names=ve) #Crear el gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('ArbolDecisionCollegeAdmisions.png')

True

**Descripción de los datos:**

De acuerdo con la base de datos de College Admisions, esta posee un total de 17339 datos totales divididos en 1303 admitidos y 225 no admitidos. Cada uno de los aplicantes estará descrito en función de 7 variables socieconómicas:

•	Edu_Parent1 y Edu_Parent2: Nivel educativo alcanzado por cada uno de los padres o tutores del estudiante.

•	Gender: Género del estudiante (Masculino o Femenino).

•	White: Variable que indica si un estudiante es identificado como blanco en el contexto social.

•	Asian: Variable que indica si un estudiante es identificado como asiático.

•	HSGPA (High School GPA): Variable que indica el promedio de calificaciones del estudiante en la secundaria en una escala de 0 a 5.

•	SAT/ACT: Variable que indica el puntaje obtenido en los exámenes estandarizados SAT o ACT en EEUU.









3. Se procede a saber el número de personas Admitidas y No Admitidas

In [ ]:
# Asumiendo que 'XDB' es tu DataFrame y 'Admitted' es la columna que indica la admisión
admitidos_counts = XDB['Admitted'].value_counts()

# Imprimir los resultados
print(admitidos_counts)

Admitted
Yes    1303
No      225
Name: count, dtype: int64


4. Evaluar el comportamiento del modelo

In [ ]:
ydp=mar.predict(XD) #Pronóstico sobre la base de datos original
cm=confusion_matrix(yd, ydp)
print("La matriz de confusión es: \n", cm)

VN=cm[0,0]; FP=cm[0,1];FN=cm[1,0]; VP=cm[1,1]

Exactitud=(VP+VN)/(VP+VN+FP+FN)
Sensibilidad=VP/(VP+FN)
Especificidad=VN/(VN+FP)
Precision=VP/(VP+FP)
TasaError=(FP+FN)/(VP+VN+FP+FN)
PredNeg=VN/(VN+FN)

print("La exactitud es: ", Exactitud)
print("La sensibilidad es: ", Sensibilidad)
print("La especificidad es: ", Especificidad)
print("La precision es: ", Precision)
print("La tasa de error es: ", TasaError)
print("La presición negativa es:", PredNeg)

La matriz de confusión es: 
 [[  20  205]
 [   2 1301]]
La exactitud es:  0.8645287958115183
La sensibilidad es:  0.9984650805832693
La especificidad es:  0.08888888888888889
La precision es:  0.8638778220451527
La tasa de error es:  0.1354712041884817
La presición negativa es: 0.9090909090909091


**Análisis de resultados**

La matriz de confusión muestra una predominancia por los valores Verdaderos Positivos (1301) y los valores Falsos Positivos (205), esto hizo que el modelo alcanzara una alta sensibilidad (99.84%) y una buena precisión (86.38%), lo que indica un buen desempeño en la identificación de los casos positivos.
Con respecto a la clasificación de los casos negativos, la base de datos mostró un total de 225 muestras negativas, de las cuales 20 fueron correctamente clasificadas como negativas y 205 fueron clasificadas incorrectamente como positivas. Se destaca que la especificidad fue considerablemente baja (8.89%), lo que indica que el modelo tiene dificultades para identificar correctamente los casos negativos, además, en el análisis del modelo se encontraron 7 nodos puros, lo que sugiere que ciertas reglas de decisión están bien definidas en algunos segmentos del conjunto de datos.

De acuerdo con los índices obtenidos, el modelo muestra un mejor comportamiento frente a la sensibilidad, mientras que la especificidad es baja en comparación, por lo cual, es importante destacar la necesidad de ajustar parámetros o estrategias para mejorar la clasificación de los casos negativos sin afectar significativamente la capacidad de detección de los positivos.


4. Informe en excel

In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

# ... (código anterior para cargar datos y entrenar el modelo) ...

# 1. Cargar la hoja 1 del archivo de Excel
nxl = '/content/drive/MyDrive/2. CollegeAdmisions_Int_M.xlsx'
hoja1_df = pd.read_excel(nxl, sheet_name=1)  # Cargar la hoja 1

# 2. Seleccionar las variables para el pronóstico
XD_hoja1 = hoja1_df[['Edu_Parent1', 'Edu_Parent2', 'Gender', 'White', 'Asian', 'HSGPA', 'SAT/ACT']]
XD_hoja1.loc[:, 'Gender'] = XD_hoja1['Gender'].map({'F': 0, 'M': 1})

# 3. Realizar el pronóstico
pronóstico_hoja1 = mar.predict(XD_hoja1)

# 4. Agregar el pronóstico al DataFrame de la hoja 1
hoja1_df['Pronóstico'] = pronóstico_hoja1

# 5. Guardar el DataFrame con el pronóstico en un nuevo archivo de Excel
hoja1_df.to_excel('Pronóstico_Admisiones_Hoja1.xlsx', index=False)

print("Pronóstico generado y guardado en 'Pronóstico_Admisiones_Hoja1.xlsx'")

Pronóstico generado y guardado en 'Pronóstico_Admisiones_Hoja1.xlsx'
